app.go

```diff
@@ -92,241 +146,132 @@ type App struct {
 	serviceDiscovery cluster.ServiceDiscovery
 	startAt          time.Time
 	worker           *worker.Worker
-}
-
-var (
-	app = &App{
-		server:           cluster.NewServer(uuid.New().String(), "game", true, map[string]string{}),
+	remoteService    *service.RemoteService
+	handlerService   *service.HandlerService
+	handlerComp      []regComp
+	remoteComp       []regComp
+	modulesMap       map[string]interfaces.Module
+	modulesArr       []moduleWrapper
+	groups           groups.GroupService
+	sessionPool      session.SessionPool
+}
+
+// NewApp is the base constructor for a pitaya app instance
+func NewApp(
+	serverMode ServerMode,
+	serializer serialize.Serializer,
+	acceptors []acceptor.Acceptor,
+	dieChan chan bool,
+	router *router.Router,
+	server *cluster.Server,
+	rpcClient cluster.RPCClient,
+	rpcServer cluster.RPCServer,
+	worker *worker.Worker,
+	serviceDiscovery cluster.ServiceDiscovery,
+	remoteService *service.RemoteService,
```

Comment:
Why are you passing this strcut on the constructor instead of creating it internally how it was before?

Reply:
The idea is to avoid deep construction hierarchies. The deeper they are, the harder it is to refactor stuff. Also, it's easier to mock dependencies.

The alternative would be to access private fields or create public getters just for this purpose. And this last one would make possible to change it after it's creation, which could cause unpredictable behaviours and more edge cases to test.